pydantic_ai_graph/pydantic_ai_graph/graph.py

```diff
@@ -2,169 +2,232 @@
+            if len(bad_edges_list) == 1:
+                raise ValueError(f'{bad_edges_list[0]} but not included in the graph.')
+            else:
+                b = '\n'.join(f' {be}' for be in bad_edges_list)
+                raise ValueError(f'Nodes are referenced in the graph but not included in the graph:\n{b}')
 
     async def run(
+        self, state: StateT, node: BaseNode[StateT, RunEndT]
+    ) -> tuple[RunEndT, list[GraphHistoryItem[StateT, RunEndT]]]:
+        if not isinstance(node, self.nodes):
+            raise ValueError(f'Node "{node.get_id()}" is not in the graph.')
+        run = GraphRun[StateT, RunEndT](state=state)
+        result = await run.run(self.name, node)
+        history = run.history
+        return result, history
+
+    def get_runner(
         self,
-        input_data: GraphInputT,
-        state: StateT = None,
-        history: list[Snapshot] | None = None,
-    ) -> tuple[GraphOutputT, list[Snapshot]]:
-        current_node_def = self.first_node
-        current_node = current_node_def.node(input_data)
-        ctx = GraphContext(state)
-        if history:
-            run_history = history[:]
-        else:
-            run_history = []
+        first_node: StartNodeProtocol[RunSignatureT, StateT, RunEndT],
+    ) -> GraphRunner[RunSignatureT, StateT, RunEndT]:
+        return GraphRunner(
+            graph=self,
+            first_node=first_node,
+        )
+
+
+@dataclass
+class GraphRunner(Generic[RunSignatureT, StateT, RunEndT]):
+    """Runner for a graph.
+
+    This is a separate class from Graph so that you can get a type-safe runner from a graph definition
+    without needing to manually annotate the paramspec of the start node.
+    """
+
+    graph: Graph[StateT, RunEndT]
+    first_node: StartNodeProtocol[RunSignatureT, StateT, RunEndT]
+
+    def __post_init__(self):
+        if self.first_node not in self.graph.nodes:
+            raise ValueError(f'Start node "{self.first_node.get_id()}" is not in the graph.')
+
+    async def run(
+        self, state: StateT, /, *args: RunSignatureT.args, **kwargs: RunSignatureT.kwargs
+    ) -> tuple[RunEndT, list[GraphHistoryItem[StateT, RunEndT]]]:
+        run = GraphRun[StateT, RunEndT](state=state)
+        result = await run.run(self.graph.name, self.first_node(*args, **kwargs))
+        history = run.history
+        return result, history
+
+    def mermaid_code(self) -> str:
+        return mermaid_code(self.graph, self.first_node)
+
+    def mermaid_image(self, mermaid_ink_params: dict[str, str | int] | None = None) -> bytes:
+        return mermaid_image(self.graph, self.first_node, mermaid_ink_params)
+
+    def mermaid_save(self, path: Path | str, mermaid_ink_params: dict[str, str | int] | None = None) -> None:
+        mermaid_save(path, self.graph, self.first_node, mermaid_ink_params)
+
+
+@dataclass
+class GraphRun(Generic[StateT, RunEndT]):
```

Comment:
are we sure we need this? Maybe we could make it private if we're not sure?

Reply:
It's currently the only thing that exposes the `step` API, which I feel like you want for unit testing. Note that I think we want to expose a way to do unit testing that goes beyond just calling `node.run` specifically for the sake of testing callbacks (once we add support for them).

But I think we could rework so `GraphRun` is private and there is a public function that users can call that basically does the same thing as `pydantic_ai_graph.graph.GraphRun.step` with callback execution, but doesn't require access to that class.

Reply:
To be clear I definitely don't think we need to have this class, it could be replaced with function calls operating on `state` and `history` directly, but:
* My comment above was a response to the suggestion of making it private — I think the functionality for stepping through a graph (with callbacks) should be public. (But that doesn't mean it needs to be a class.)
* I think having a class will make it easier to add features over time (in particular callbacks, though maybe other things). But we could drop the class for now and/or make it private as a way to keep users from depending on APIs we may want to break in the near future.